frontera/contrib/backends/redis_backend/__init__.py

```diff
@@ -0,0 +1,387 @@
+# -*- coding: utf-8 -*-
+from collections import Iterable
+from datetime import datetime
+from frontera.utils.url import parse_domain_from_url_fast
+from frontera import DistributedBackend
+from frontera.core.components import Metadata, Queue, States
+from frontera.core.models import Request
+from frontera.contrib.backends.partitioners import Crc32NamePartitioner
+from frontera.utils.misc import get_crc32
+from frontera.contrib.backends.remote.codecs.msgpack import Decoder, Encoder
+import logging
+from msgpack import packb, unpackb
+from redis import ConnectionPool, StrictRedis
+from time import time
+
+FIELD_CRAWL_AT = b'crawl_at'
+FIELD_CREATED_AT = b'created_at'
+FIELD_DEPTH = b'depth'
+FIELD_DOMAIN = b'domain'
+FIELD_DOMAIN_FINGERPRINT = b'domain_fingerprint'
+FIELD_ERROR = b'error'
+FIELD_FINGERPRINT = b'fingerprint'
+FIELD_NAME = b'name'
+FIELD_SCORE = b'score'
+FIELD_STATE = b'state'
+FIELD_STATUS_CODE = b'status_code'
+FIELD_URL = b'url'
+
+logging.getLogger('boto3.resources.action').setLevel(logging.WARNING)
+
+
+class RedisQueue(Queue):
+    MAX_SCORE = 1.0
+    MIN_SCORE = 0.0
+    SCORE_STEP = 0.01
+
+    def __init__(self, pool, partitions, delete_all_keys=False):
+        self._pool = pool
+        self._partitions = [i for i in range(0, partitions)]
+        self._partitioner = Crc32NamePartitioner(self._partitions)
+        self._logger = logging.getLogger("redis.queue")
+
+        if delete_all_keys:
+            connection = StrictRedis(connection_pool=self._pool)
+            connection.flushdb()
+
+        class DumbResponse:
+            pass
+
+        self._decoder = Decoder(Request, DumbResponse)
+        self._encoder = Encoder(Request)
+
+    def get_next_requests(self, max_n_requests, partition_id, **kwargs):
```

Comment:
Overall, what would be the use case for RedisQueue? Would it be possible to use it for large crawls (millions of websites), or only for small ones (up to 100) ? Another question, is the use case requires broad crawling capabilities? E.g. getting batch for minimum of X domains with no more than Y uniformly distributed number of pages per domain? Some backends don't have this feature implemented because they aren't expected to be used for large scale crawls.
From my POV we can't put much data in Redis, but if crawler is relatively fast and queue will not grow beyond some limits it's fine.

The same questions regarding other parts RedisStates and RedisMetadata.